### PR TITLE
Query cache: Enable compression of cache entries

### PR DESCRIPTION
**Pull request by rschu1ze**
This PR changes two things in the query cache:

- ClickHouse reads table data in blocks of `max_block_size` rows but due to filtering, aggregation, etc., result blocks are typically much smaller than `max_block_size` (and sometimes bigger). The new setting `query_cache_squash_partial_results` (enabled by default) now controls if result blocks are squashed (if they are tiny) or split (if they are large) into blocks of `max_block_size` size before insertion into the query result cache. This reduces performance of writes into the query cache but improves compressability of cache entries and provides a "natural" block granularity when query results are later served from the query cache.

- Entries in the query cache are now compressed by default. This reduces the overall memory consumption at the cost of slower writes into / reads from the query cache. To disable compression, use the new setting `query_cache_compress_entries`.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Entries in the query cache are now squashed to max_block_size and compressed

### CODE REVIEW

File: src/Interpreters/Cache/QueryCache.cpp [hunk 1]
```diff
@@ -218,6 +220,25 @@ void QueryCache::Writer::finalizeWrite()
 
     std::lock_guard lock(mutex);
 
+    auto to_single_chunk = [](const std::shared_ptr<Chunks>& chunks_) -> std::shared_ptr<Chunks>
```
Comment (alexey-milovidov) [inline src/Interpreters/Cache/QueryCache.cpp:223]: If the resultset is large, making it a single chunk can be worse. Let's just squash to max_block_size.

File: docs/en/operations/server-configuration-parameters/settings.md [hunk 2]
```diff
@@ -1279,7 +1279,10 @@ The following settings are available:
 -   `size`: The maximum cache size in bytes. 0 means the query cache is disabled. Default value: `1073741824` (1 GiB).
 -   `max_entries`: The maximum number of `SELECT` query results stored in the cache. Default value: `1024`.
 -   `max_entry_size`: The maximum size in bytes `SELECT` query results may have to be saved in the cache. Default value: `1048576` (1 MiB).
--   `max_entry_records`: The maximum number of records `SELECT` query results may have to be saved in the cache. Default value: `30000000` (30 mil).
+-   `max_entry_rows`: The maximum number of rows `SELECT` query results may have to be saved in the cache. Default value: `30000000` (30 mil).
+-   `compress_entries`: If cache entries are compressed. Reduces the cache memory consumption but reduces cache insert/read performance. Default value: `true`.
```
Comment (alexey-milovidov) [inline docs/en/operations/server-configuration-parameters/settings.md:1283]: Another idea is to have the compressed flag near each entry, and then the compression can be controlled by a query-level setting on the side of the query writing to the cache.

Comment (rschu1ze) [reply docs/en/operations/server-configuration-parameters/settings.md:1283]: Sounds good, can do.

File: src/Interpreters/Cache/QueryCache.cpp (hunk 1 above)
Comment (rschu1ze) [reply src/Interpreters/Cache/QueryCache.cpp:223]: Yepp, will check.

File: src/Interpreters/executeQuery.cpp [hunk 3]
```diff
@@ -726,7 +726,8 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
                     QueryCache::Key key(
                         ast, res.pipeline.getHeader(),
                         std::make_optional<String>(context->getUserName()),
-                        std::chrono::system_clock::now() + std::chrono::seconds(settings.query_cache_ttl));
+                        /*dummy value for expires_at*/ std::chrono::system_clock::from_time_t(1),
+                        /*dummy value for is_compressed*/ true);
```
Comment (SmitaRKulkarni) [inline src/Interpreters/executeQuery.cpp:730]: maybe we can set these as default values for function parameters

File: src/Storages/System/StorageSystemQueryCache.cpp [hunk 4]
```diff
@@ -16,6 +16,7 @@ NamesAndTypesList StorageSystemQueryCache::getNamesAndTypes()
         {"key_hash", std::make_shared<DataTypeUInt64>()},
         {"expires_at", std::make_shared<DataTypeDateTime>()},
         {"stale", std::make_shared<DataTypeUInt8>()},
+        {"compressed", std::make_shared<DataTypeUInt8>()},
```
Comment (SmitaRKulkarni) [inline src/Storages/System/StorageSystemQueryCache.cpp:19]: just curious, why we don't add this column to the end ? readability is affected after result size ?

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

File: src/Storages/System/StorageSystemQueryCache.cpp (hunk 4 above)
Comment (rschu1ze) [reply src/Storages/System/StorageSystemQueryCache.cpp:19]: Didn't think too much about the order :) But you are right, perhaps the important stuff (query, result_size, stale), should come first. I'd adjust that in a separate PR.

File: tests/queries/0_stateless/02494_query_cache_squash_partial_results.sql [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+-- Run query which reads multiple chunks (small max_block_size), cache result in query cache, but **disable** squashing of partial results
+SELECT * FROM t ORDER BY c SETTINGS max_block_size = 2, use_query_cache = true, query_cache_squash_partial_results = false;
+
+SELECT '-';
+
+-- Run again to check that no bad things happen and that the result is as expected
+SELECT * FROM t ORDER BY c SETTINGS max_block_size = 2, use_query_cache = true;
+
+DROP TABLE t;
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply tests/queries/0_stateless/02494_query_cache_squash_partial_results.sql:48]: Yepp, that makes sense. Allow me to do that in a separate PR.

[the remark this answers is not in the document]
Comment (rschu1ze) [reply tests/queries/0_stateless/02494_query_cache_squash_partial_results.sql:48]: Ah, better let me push a follow-up commit into this PR.

**Comment by CurtizJ** [comment]
See https://github.com/ClickHouse/ClickHouse/pull/48358#issuecomment-1494710268.

File: tests/queries/0_stateless/02494_query_cache_compression.sql [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+
+SET allow_experimental_query_cache = true;
+
+SYSTEM DROP QUERY CACHE;
+
+DROP TABLE IF EXISTS t;
+
+-- Create test table with lot's of rows
+CREATE TABLE t(c String) ENGINE=MergeTree ORDER BY c;
+INSERT INTO t values ('abc') ('def') ('abc') ('jkl');
```
Comment (devcrafter) [inline tests/queries/0_stateless/02494_query_cache_compression.sql:12]: Please
```sql
SELECT multiIf(n = 0, 'abc', n = 1, 'def', n = 2, 'abc', 'jkl')
FROM
(
    SELECT number % 4 AS n
    FROM numbers(4 * 400)
)
```

Comment (rschu1ze) [reply tests/queries/0_stateless/02494_query_cache_compression.sql:12]: #48373, thanks.

**Pull request merged**